src/mongo/db/pipeline/expression.cpp

```diff
@@ -2144,74 +2145,81 @@ const char* ExpressionLog10::getOpName() const {
-        Value pResult(evaluateInternal(&emptyVars));
-        intrusive_ptr<Expression> pReplacement(ExpressionConstant::create(pResult));
-        return pReplacement;
-    }
-
-    // Remaining optimizations are only for associative and commutative expressions.
-    if (!isAssociativeAndCommutative())
-        return this;
-
-    // Process vpOperand to split it into constant and nonconstant vectors.
-    // This can leave vpOperand in an invalid state that is cleaned up after the loop.
-    ExpressionVector constExprs;
-    ExpressionVector nonConstExprs;
-    for (size_t i = 0; i < vpOperand.size(); ++i) {  // NOTE: vpOperand grows in loop
-        intrusive_ptr<Expression> expr = vpOperand[i];
-        if (dynamic_cast<ExpressionConstant*>(expr.get())) {
-            constExprs.push_back(expr);
-        } else {
-            // If the child operand is the same type as this and is also associative and
-            // commutative, then we can extract its operands and inline them here. We detect
-            // sameness of the child operator by checking for equality of the opNames
-            ExpressionNary* nary = dynamic_cast<ExpressionNary*>(expr.get());
-            if (!nary || !str::equals(nary->getOpName(), getOpName()) ||
-                !nary->isAssociativeAndCommutative()) {
-                nonConstExprs.push_back(expr);
-            } else {
-                // same expression, so flatten by adding to vpOperand which
-                // will be processed later in this loop.
-                vpOperand.insert(vpOperand.end(), nary->vpOperand.begin(), nary->vpOperand.end());
-            }
+    for (auto& operand : vpOperand) {
+        operand = operand->optimize();
+        if (dynamic_cast<ExpressionConstant*>(operand.get())) {
+            ++constOperandCount;
         }
     }
-
-    // collapse all constant expressions (if any)
-    Value constValue;
-    if (!constExprs.empty()) {
-        vpOperand = constExprs;
+    if (constOperandCount == vpOperand.size()) {
         Variables emptyVars;
-        constValue = evaluateInternal(&emptyVars);
+        return intrusive_ptr<Expression>(ExpressionConstant::create(evaluateInternal(&emptyVars)));
     }
 
-    // now set the final expression list with constant (if any) at the end
-    vpOperand = nonConstExprs;
-    if (!constExprs.empty()) {
-        vpOperand.push_back(ExpressionConstant::create(constValue));
+    if (isAssociative()) {
+        ExpressionVector constExpressions;
+        ExpressionVector optimizedOperands;
+        for (size_t i = 0; i < vpOperand.size();) {
+            intrusive_ptr<Expression> operand = vpOperand[i];
+            if (dynamic_cast<ExpressionConstant*>(operand.get())) {
+                constExpressions.push_back(operand);
+                ++i;
+                continue;
+            }
+            ExpressionNary* nary = dynamic_cast<ExpressionNary*>(operand.get());
+            if (nary && str::equals(nary->getOpName(), getOpName())) {
```

Comment:
The check of `nary->isAssociative()` is still needed. It's subtle, but see [this commit](https://github.com/mongodb/mongo/commit/f9828f5a0312801fa5b8592b31648a64f0ea7f67) for why it's necessary, particularly [this test and comment](https://github.com/mongodb/mongo/commit/f9828f5a0312801fa5b8592b31648a64f0ea7f67#diff-22feeb9ade901bc4e660448dcbf9fe61R69).

Unfortunately, I don't think that test ever caught the bug it was intended to catch, since `optimize()` is called first, and the array of operands will be reduced to one constant. If you add a "$a" to the nested expression, it should not optimize to a constant, so the test should fail. Can you please update the test, make sure it fails for your current set of changes, then add back the `nary->isAssociative()` check to ensure it is fixed?

Let me know if that's unclear, as it is very subtle.